docs/application/web/guides/machine-learning/pipeline.md

```diff
@@ -0,0 +1,445 @@
+   ```javascript
+   insSwitch.select('sink_1'); // green branch used as the source
+   ```
+
+In the example above, we used an `input-selector` to choose from alternative sources.
+Similarily, we can use an `output-selector` to direct data to one of a set of sinks.
+
+
+## Start and stop data flow to a pipeline branch
+Use `valve` element with `Valve` API to start and stop data flow in one of a pipeline branches.
+
+To control data flow in a single pipeline branch:
+1. Create and start a pipeline with a `valve` element:
+   ```javascript
+   var pipelineDescription = 'videotestsrc is-live=true ' +
+                             '! videoconvert ' +
+                             '! videoscale ' +
+                             '! video/x-raw,format=RGBx,width=16,height=16,framerate=10/1 ' +
+                             '! tensor_converter ' +
+                             '! valve name=valve1 ' +
+                             '! fakesink';
+
+   var pipeline = tizen.ml.pipeline.createPipeline(pipelineDescription);
+   pipeline.start();
+   ```
+
+2. Get `Valve` element from the pipeline:
+   ```javascript
+   var valve = pipeline.getValve('valve1');
+   ```
+
+3. You can check if the valve is opened or closed:
+   ```javascript
+   console.log(valve.isOpen); // true
+   ```
+
+4. Set the desired valve state:
+   ```javascript
+   valve.setOpen(false); // stop the data flow
+   ```
+
+
+## Write custom data filters
+The data flowing through the pipeline can be transformed in application, using  `CustomFilter` callbacks.
+
+   > [!NOTE]
+   > Sending data between `CustomFilter` running in JS application and GStreamer pipeline has large overhead.
+   > Using `CustomFilter` with large tensors may result in unsatisfactory performance.
+
+To transform the data within a callback registered in JS application:
+1. Define information about `CustomFilter`'s input and output tensors:
+   ```javascript
+   var inputInfo = new tizen.ml.TensorsInfo();
+   inputInfo.addTensorInfo('3D', 'UINT8', [4, 20, 15, 1]);
+   var outputInfo = new tizen.ml.TensorsInfo();
+   outputInfo.addTensorInfo('flat', 'UINT8', [1200]);
+   ```
+
+2. Register a `CustomFilter` in the pipeline. Remember about returning a proper status code from the callback - in this case we return `0` to indicate success:
+   ```javascript
+   function flattenFilter(inputData, outputData) {
+       var rawInputData = inputData.getTensorRawData(0);
+       outputData.setTensorRawData(0, rawInputData.data);
+
+       return 0;
+   };
+
+   tizen.ml.pipeline.registerCustomFilter('flattenFilter', flattenFilter, inputInfo, outputInfo);
+   ```
+
+   > [!Note]
+   > For performance reasons `inputData` and `outputData` passed to `flattenFilter` behave differently than ordinary `TensorsData` objects.
+   > They cannot be disposed and are invalidated when the callback returns - you have to copy them manually to use outside the callback.
```

Comment:
The sentences are not clear in note, please rephrase. 
> [!NOTE]
   > For performance reasons, `inputData` and `outputData` are passed to `flattenFilter`,  behave differently than ordinary `TensorsData` objects.
   > They cannot be disposed and are invalidated when the callback returns - you have to copy them manually to use outside the callback.